### PR TITLE
Add support for tmpfs and KVDB rparams to disable O_DIRECT (NFSE-4771)

### PR DESCRIPTION
**Pull request by nabeelmmd**
Signed-off-by: Nabeel M Mohamed <50154757+nabeelmmd@users.noreply.github.com>

### CODE REVIEW

File: lib/kvdb/kvdb_rparams.c [hunk 1]
```diff
@@ -1433,6 +1433,51 @@ static const struct param_spec pspecs[] = {
             }
         },
     },
+    {
+        .ps_name = "storage.capacity.dio_disable",
```
Comment (tristan957) [inline lib/kvdb/kvdb_rparams.c:1437]: Change to storage.capacity.direct_io.enabled.

More verbose and matches the previous "enabled" "disabled" terminology we have going on

File: lib/mpool/src/mpool.c [hunk 2]
```diff
@@ -7,6 +7,11 @@
 
 #include <limits.h>
 #include <bsd/string.h>
+#include <sys/vfs.h>
+
+#if __linux__
+#include <linux/magic.h>
```
Comment (tristan957) [inline lib/mpool/src/mpool.c:13]: Magical :)

File: lib/mpool/src/mpool.c [hunk 3]
```diff
@@ -105,10 +110,44 @@ mclass_path_check(enum hse_mclass mclass, const char *path)
...
+/*
+ * This function currently uses a Linux-specific statfs(2) syscall.
+ *
+ * A portable way to implement this will be to traverse all mounted file-systems using
+ * getmntent() and check whether the device ID ('st_dev' from stat(2)) of any of the
+ * tmpfs file-systems matches with the device ID of 'path'.
+ *
+ * At this point, using statfs(2) is lightweight and this can be revisited when we
+ * port HSE to other platforms.
+ *
```
Comment (tristan957) [inline lib/mpool/src/mpool.c:122]: Delete this line.

File: lib/mpool/src/mpool.c [hunk 4]
```diff
@@ -308,12 +347,27 @@ mpool_open(
...
+            bool tmpfs;
+
+            err = mclass_path_is_tmpfs(mcp.path, &tmpfs);
+            if (err)
+                goto errout;
+
+            if (!tmpfs)
+                oflags |= O_DIRECT;
+            else
+                log_info("Disabling O_DIRECT access as the mclass(%d) path(%s) is on a tmpfs",
```
Comment (tristan957) [inline lib/mpool/src/mpool.c:366]: Put a space before the () stuff

File: lib/mpool/src/mpool_file.c [hunk 5]
```diff
@@ -61,6 +61,10 @@ mpool_file_open(
     if (create)
         flags |= (O_CREAT | O_EXCL);
 
+    /* Drop ODIRECT if the mclass doesn't support direct-io */
+    if ((flags & O_DIRECT) && !mclass_directio(mc))
```
Comment (tristan957) [inline lib/mpool/src/mpool_file.c:65]: Maybe name the function `mclass_is_direct_io_enabled()`?

Note that I split `directio` to `direct_io` based on the earlier parameter comments

Comment (tristan957) [reply lib/mpool/src/mpool_file.c:65]: Kind of a verbose name, but gets the point across easily.

**Review by tristan957 — APPROVED** [review]

File: lib/mpool/src/mpool.c (hunk 2 above)
Comment (tristan957) [reply lib/mpool/src/mpool.c:13]: What is in this header? Never heard of it.

File: lib/kvdb/kvdb_rparams.c (hunk 1 above)
Comment (nabeelmmd) [reply lib/kvdb/kvdb_rparams.c:1437]: Yes, I like that! Will change it to storage.capacity.directio.enabled and default it to true.

File: lib/mpool/src/mpool.c (hunk 2 above)
Comment (nabeelmmd) [reply lib/mpool/src/mpool.c:13]: It contains the superblock magic of the filesystems in Linux, I think...

File: lib/mpool/src/mpool_file.c (hunk 5 above)
Comment (nabeelmmd) [reply lib/mpool/src/mpool_file.c:65]: Yeah, it is verbose for an internal function. I can do mclass_directio_get(mc).

File: lib/include/hse_ikvdb/kvdb_rparams.h [hunk 6]
```diff
@@ -101,7 +101,7 @@ struct kvdb_rparams {
 
     uint32_t keylock_tables;
 
-    bool   dio_disable[HSE_MCLASS_COUNT];
+    bool   dio_enable[HSE_MCLASS_COUNT];
```
Comment (tristan957) [inline lib/include/hse_ikvdb/kvdb_rparams.h:104]: Last nit...dio_enabled?

Comment (nabeelmmd) [reply lib/include/hse_ikvdb/kvdb_rparams.h:104]: Right, I had the same thought. Then I saw perfc_enable, dur_enable etc. in this struct and went with dio_enable.

Comment (nabeelmmd) [reply lib/include/hse_ikvdb/kvdb_rparams.h:104]: Actually, let me change it to dio_enabled. That sounds correct.

Comment (nabeelmmd) [reply lib/include/hse_ikvdb/kvdb_rparams.h:104]: leaving this unchanged for now...

File: lib/kvdb/kvdb_rparams.c (hunk 1 above)
Comment (beckerg) [reply lib/kvdb/kvdb_rparams.c]: Unfortunate, because "enable/disable" are directives, while "enabled/disabled" are observations of current state...

Comment (nabeelmmd) [reply lib/kvdb/kvdb_rparams.c]: You mean enable/disable is better than enabled/disabled for these rparams, right?

Comment (alexttx) [reply lib/kvdb/kvdb_rparams.c]: In the old days, we tried to have `0` be the default value for all params (all params were scalar types).  So some were named `enable`, others were named `disable` depending on the desired default behavior.

File: lib/mpool/include/mpool/mpool_structs.h [hunk 7]
```diff
@@ -52,6 +52,7 @@ struct mpool_cparams {
  */
 struct mpool_rparams {
     struct {
+        bool dio_disable;
```
Comment (beckerg) [inline lib/mpool/include/mpool/mpool_structs.h:55]: Yikes, this is like 4K per mclass...

File: lib/mpool/src/mpool_file.c [hunk 8]
```diff
@@ -61,6 +61,10 @@ mpool_file_open(
...
+        flags &= ~O_DIRECT;
+
```
Comment (beckerg) [inline lib/mpool/src/mpool_file.c:67]: s/ODIRECT/O_DIRECT/

**Review by beckerg — APPROVED** [review]

File: lib/mpool/src/mpool_file.c (hunk 8 above)
Comment (nabeelmmd) [reply lib/mpool/src/mpool_file.c:67]: Greg, please refresh. I fixed this in the last update almost an hour back..

File: lib/mpool/include/mpool/mpool_structs.h (hunk 7 above)
Comment (tristan957) [reply lib/mpool/include/mpool/mpool_structs.h:55]: I guess these buffers could converted to flexible arrays?

Comment (beckerg) [reply lib/mpool/include/mpool/mpool_structs.h:55]: Maybe not a big deal, I was thinking they were stored in an rparams structure somewhere, but I'm not seeing that...

Comment (nabeelmmd) [reply lib/mpool/include/mpool/mpool_structs.h:55]: Yes, struct mpool_c/r/dparams are not embedded in any long-lived structs.

At kvdb creation, these paths are serialized into the kvdb.meta file. They are de-serialized from this file at runtime on a need basis.

**Pull request merged**